src/web/middleware/precondition.rs

```diff
@@ -76,140 +76,123 @@ where
-                    ),
-                ));
+        async move {
+            if DOCKER_FLOW_ENDPOINTS.contains(&sreq.uri().path().to_lowercase().as_str()) {
+                return service.call(sreq).await;
             }
-        };
-        let user_id = match sreq.get_hawk_id() {
-            Ok(v) => v,
-            Err(e) => {
-                warn!("⚠️ Hawk header error {:?}", e);
-                queue_report(sreq.extensions_mut(), &e);
-                return Box::pin(future::ok(
-                    sreq.into_response(
+
+            // Pre check
+            let tags = {
+                let exts = sreq.extensions();
+                match exts.get::<Tags>() {
+                    Some(t) => t.clone(),
+                    None => Tags::from_request_head(sreq.head()),
+                }
+            };
+            let precondition =
+                match PreConditionHeaderOpt::extrude(&sreq.headers(), Some(tags.clone())) {
+                    Ok(precond) => match precond.opt {
+                        Some(p) => p,
+                        None => PreConditionHeader::NoHeader,
+                    },
+                    Err(e) => {
+                        warn!("⚠️ Precondition error {:?}", e);
+                        queue_report(sreq.extensions_mut(), &e);
+                        return Ok(sreq.into_response(
+                            HttpResponse::BadRequest()
+                                .content_type("application/json")
+                                .body("An error occurred in preprocessing".to_owned())
+                                .into_body(),
+                        ));
+                    }
+                };
+            let user_id = match sreq.get_hawk_id() {
+                Ok(v) => v,
+                Err(e) => {
+                    warn!("⚠️ Hawk header error {:?}", e);
+                    queue_report(sreq.extensions_mut(), &e);
+                    return Ok(sreq.into_response(
                         HttpResponse::Unauthorized()
                             .content_type("application/json")
                             .body("Invalid Authorization".to_owned())
                             .into_body(),
-                    ),
-                ));
-            }
-        };
-        let edb = extrude_db(&sreq.extensions());
-        let db = match edb {
-            Ok(v) => v,
-            Err(e) => {
-                error!("⚠️ Database access error {:?}", e);
-                queue_report(sreq.extensions_mut(), &e);
-                return Box::pin(future::ok(
-                    sreq.into_response(
+                    ));
+                }
+            };
+
+            let (req, payload) = sreq.into_parts();
+            let db_pool = extrude_db_pool(&req).await?;
+            let edb = db_pool.get().await.map_err(Error::from);
+            let sreq = ServiceRequest::from_parts(req, payload)
+                .unwrap_or_else(|_| panic!("TODO: this should never happen"));
```

Comment:
This is a bit iffy, probably better to manually extract the DB pool from state. `ServiceRequest::from_parts` panics if there is more than one copy of the request (shouldn't happen with this code, but is a liability).